### ISSUE TITLE
Mixpanel messaging support .

### ISSUE THREAD

**Issue by scdyn** [samzilverberg/cordova-mixpanel-plugin#91]
Besides event tracking and analytics, Mixpanel allows for InApp messaging and push notifications.

The entry point of their SDK to such features are MixpanelAPI.getPeople().getNotificationIfAvailable() and .initPushHandling()

AFAIK, the cordova-mixpanel-plugin doesn't support these, meaning that the plugin doesn't support messaging yet.
Am I wrong ? Any plan regarding messaging support ?

**Comment by samzilverberg** [comment]
hi @scdyn , sry for the delayed message. I missed the email notification for this issue.
You're right that the plugin doesn't support this yet.

To let the app receive notifs, for android for example, a `GCMReceiver` needs to be defined in the android manifest ([see example here](https://mixpanel.com/help/reference/android-push-notifications).

I deliberately avoided implementing this because this doesn't play well nice with the more common way of handling push notifs aka PhoneGap Push Plugin.
I don't remember exactly the details, but i remember either both receivers get the notifs or just one of them, causing unexpected behavior.

cordova and plugin.xml have evolved a bit since a couple of years ago so maybe we can find a nice conditional way to implement this and make it optional via some plugin installation flag.

I'll try looking into it but dont see it being in the codebase anytime in next month or two.


other than that, I checked the mixpanel reference for in-app messages at:

- ios: https://mixpanel.com/help/reference/ios-inapp-messages
- android: https://mixpanel.com/help/reference/android-inapp-messages

looks like ios&android work a little differently, which would mean more effort needed to integrate this...

**Comment by SpellChucker** [comment]
FYI, iPhone works great with both in-app and push notifications here in conjunction with the PhoneGap Push Plugin. I am having issues with Android and I think it might be related to not calling `initPushHandling`. We don't have a way of calling that via this plugin. Maybe we can add that method? It takes the `SENDER_ID` as a parameter.

Other than that, it should just be setting up the permissions and such but I have not tried it yet. I will come back here if I have any success.

**Comment by SpellChucker** [comment]
I was able to get push notifications working for android by adding in the required configuration into `AndroidManifest.xml` as well as calling `initPushHandling` for Android. I've created [this](https://github.com/samzilverberg/cordova-mixpanel-plugin/pull/94) PR to allow users to call `initPushHandling` whenever they need (should be called after `identify` as noted in the [docs](https://mixpanel.com/help/reference/android-push-notifications#letting)).

**Comment by tamsel12** [comment]
@samzilverberg  i am trying to do inappMessages. But not getting messages via mixpanel inappmessage. Could you please clarify Mixpanel Plugin do not support InAppMessages?

**Comment by FELS-Zak** [comment]
If anyone has gotten this to work for push messaging on iOS and Android, could you please tell me how?

**Comment by etiennea** [comment]
Messaging support would be great! Even if it's just the ability to call a callback handler for the message features or the ability to track opened messages manually.

**Comment by tamsel12** [comment]
Hai @samzilverberg  I am able to get InApp messages with delay. But i am not able to receive Push notifications. Could you please let me know Pushnotifications will work with Mixpanel cordova?

**Comment by samzilverberg** [comment]
yes push notifications work.
if they are not working for you, its probably a configuration issue or somethign to do with your setup/plugins combo.
either way, I can't help you with that

**Comment by tamsel12** [comment]
Yes @samzilverberg , its now working for me , But after logout the application i am not able to clear the pushtoken.  I have searched the plugin that it does not provide remove the pushtoken  from mixpanel. Is there any function to clear that push token in mixpanel?

**Comment by samzilverberg** [comment]
@tamsel12 your'e right.
the android & iphone sdks have methods to clear push tokens but the plugin doesn't expose them (yet).
no one ever asked for this before 🤷‍♂️ 

i'll see if i can add these fns to the plugin soon.

**Comment by etiennea** [comment]
To support some of the mp messaging features I did this. If useful to someone.
`
 window.phonegapPush.on('notification', data => {
            console.log('notification', data);
            let url = '';
            if (
                data.additionalData &&
                data.additionalData.mp_ontap &&
                data.additionalData.mp_ontap.type === 'deeplink' &&
                data.additionalData.mp_ontap.uri
            ) {
                url = data.additionalData.mp_ontap.uri.replace('myapp:///', '/');
            }
            if (
                data.additionalData &&
                data.additionalData.mp_ontap &&
                data.additionalData.mp_ontap.type === 'browser' &&
                data.additionalData.mp_ontap.uri
            ) {
                url = false;
                window.open(data.additionalData.mp_ontap.uri, '_system');
            }
           if (url) router.navigateTo(url);
});
`
